modal/_asgi.py

```diff
@@ -4,18 +4,43 @@
 
 from asgiref.wsgi import WsgiToAsgi
 
+from modal_proto import api_pb2
 from modal_utils.async_utils import TaskContext
+from modal_utils.grpc_utils import unary_stream
 
-from ._blob_utils import MAX_OBJECT_SIZE_BYTES
+from ._blob_utils import MAX_OBJECT_SIZE_BYTES, blob_download
+from ._serialization import deserialize_data_format
+from .client import Client
+from .functions import current_function_call_id
 
 
-def asgi_app_wrapper(asgi_app):
-    async def fn(scope, body=None):
+def asgi_app_wrapper(asgi_app, client: Client):
+    async def fn(scope):
+        function_call_id = current_function_call_id()
+        assert function_call_id, "internal error: function_call_id not set in asgi_app() scope"
+
         messages_from_app: asyncio.Queue[Dict[str, Any]] = asyncio.Queue(1)
 
         # TODO: send disconnect at some point.
         messages_to_app: asyncio.Queue[Dict[str, Any]] = asyncio.Queue(1)
-        await messages_to_app.put({"type": "http.request", "body": body})
+
+        async def fetch_inputs():
+            last_index = 0
+            while True:
+                req = api_pb2.FunctionCallGetDataRequest(function_call_id=function_call_id, last_index=last_index)
+                try:
+                    async for chunk in unary_stream(client.stub.FunctionCallGetDataIn, req):
+                        if chunk.index <= last_index:
+                            continue
+                        last_index = chunk.index
+                        if chunk.data_blob_id:
+                            message_bytes = await blob_download(chunk.data_blob_id, client.stub)
+                        else:
+                            message_bytes = chunk.data
+                        message = deserialize_data_format(message_bytes, chunk.data_format, client)
+                        await messages_to_app.put(message)
+                except Exception:  # TODO: Catch specific exceptions versus transient errors.
+                    await asyncio.sleep(0.1)
```

Comment:
This can probably be moved into a helper function in the future with shared logic if, e.g., expose the data channel for users?